### DISCUSSION TITLE
Connecting to AWSIoT

### DISCUSSION THREAD

**Discussion by FranzBusch** [adam-fowler/mqtt-nio#24] [Q&A]
Hello,

I just came across this project and was super happy that it supported AWSIoT through SOTO. We have been using the official AWSIoT iOS SDK and saw a lot of crashes and wanted to try this out as a replacement.

Sadly, when trying to connect we run into a `MQTTNIO.MQTTClient.Error.websocketUpgradeFailed)` error. I tried to get the signed url and run it with cURL however with that one I get a `404 Not found` response. I think the problem is pretty simple but I just can't figure out where it lies.

The code I use to setup the connection:

```swift
                let host = "XXX.iot.eu-west-1.amazonaws.com"
                let port = 443
                let headers = HTTPHeaders([("host", host)])
                let signer = AWSSigner(
                    credentials: StaticCredential(accessKeyId: token.accessKeyId, secretAccessKey: token.secretAccessKey, sessionToken: token.sessionToken),
                    name: "iotdata",
                    region: "eu-west-1"
                )
                let signedURL = signer.signURL(
                    url: URL(string: "https://\(host):\(port)/mqtt")!,
                    method: .GET,
                    headers: headers,
                    body: .none,
                    expires: .minutes(30)
                )
                
                let requestURI = "/mqtt?\(signedURL.query!)"
                let client = MQTTClient(
                    host: host,
                    port: port,
                    identifier: "MyAWSClient",
                    eventLoopGroupProvider: .createNew,
                    configuration: .init(useSSL: true, useWebSockets: true, webSocketURLPath: requestURI)
                )
                client.connect()
                    .whenComplete { result in
                        print(result)
                    }
```

**Comment by adam-fowler** [comment]
From first glance the code looks correct. I'll verify if I can get the code working and get back to you.

Stupid questions but they need to be asked
Do your credentials have access to IoT? 
Is eu-west-1 the correct region?

**Reply by FranzBusch** [reply]
The exact same credentials work with the AWSIoT SDK and the eu-west1 region is also correct. The only difference from a setup perspective between the two is that I am using `https://XXX.iot.eu-west-1.amazonaws.com` for the AWSIoT SDK and `XXX.iot.eu-west-1.amazonaws.com:433` for the Soto Signer.

**Reply by adam-fowler** [reply]
Can you try it without the port

**Reply by FranzBusch** [reply]
I removed the port from the `signURL` and/or the init of `MQTTClient` call I get a `failure(MQTTNIO.MQTTClient.Error.serverClosedConnection)` error.

**Reply by FranzBusch** [reply]
What I find interesting is that if I look into the network requests that our app is doing I see that only this library is trying to do the web socket init request and the AWSIoT SDK is not doing these requests.

**Reply by FranzBusch** [reply]
@adam-fowler From the docs I found `Clients must send the Server Name Indication (SNI) TLS extension` could that be something that I am missing? https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html

**Reply by adam-fowler** [reply]
What platform are you running on? And are you providing a TLS configuration or an EventLoopGroup during initialisation

**Reply by FranzBusch** [reply]
I am running on iOS in my iOS simulator. I am not passing anything right now I am just using the default `.createNew` option like in the code above.

**Reply by adam-fowler** [reply]
Did you get any further with this? I'm struggling to reproduce the issue. I can get the connect to fail if the policy assigned to my user doesn't allow for me to connect. But the websocket receives a `connectionClose` in `WebSocketHandler.channelRead` which you are saying is not the way you are see it fail. 

Could it be you are behind a proxy or something like that?

**Reply by adam-fowler** [reply]
Hi @FranzBusch, I know this is 3 months later, but I was resurrecting a MQTT NIO AWS connection sample recently and started to receive the same issue as you describe above. I managed to fix it by making sure my `identifier` had been added to the AWS policy attached to the user whose credentials you are using
```
        {
            "Effect": "Allow",
            "Action": "iot:Connect",
            "Resource": "arn:aws:iot:eu-west-1:931875313149:client/MyAWSClient"
        },
```
